### ISSUE TITLE
[Bug] [iOS] NRE in native projects (DidAppear method)

### ISSUE THREAD

**Issue by conceptdev** [xamarin/Xamarin.Forms#9269]
### Description

`NullReferenceException` - `Application` is null. Line of code referenced is **Xamarin.Forms.Platform.iOS/Platform.cs** [line 231 (line 238 on master)](https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Platform.iOS/Platform.cs#L238)

```csharp
internal void DidAppear()
{
   _animateModals = false;
   Application.Current.NavigationProxy.Inner = this;  // NRE!!!!
   _animateModals = true;
}
```

Blame for the line references [this PR](https://github.com/xamarin/Xamarin.Forms/commit/53e1d99a751e7a9cb8ae3f67ac8f3630b2aea398) which mentions an NRE. Wonder if @hartez will remember 🤔

### Steps to Reproduce

1. Run the **native2forms** sample from the xamarin-forms-samples GitHub repo
   https://docs.microsoft.com/en-us/samples/xamarin/xamarin-forms-samples/native2forms/
2. Press the **+** button to add an item
3. Press **Save** to return to the list page - NRE occurs while the list page is appearing.

### Expected Behavior

List page re-appears and refreshes with new data.

### Actual Behavior

NRE in the iOS native method `DidAppear`

### Basic Information

- Version with issue: 4.4.0.991265
- Last known good version:
- IDE: VSMac
- Platform Target Frameworks: 
  - iOS:  13.2
  - Android: n/a
- Android Support Library Version: n/a
- Nuget Packages: no others
- Affected Devices: iOS Simulators running iOS 13.2 and 13.3

### Reproduction Link

The Xamarin sample [Native2Forms](https://github.com/xamarin/xamarin-forms-samples/blob/master/Native2Forms/Notes.iOS/Views/NotesPage.xaml.cs#L19). I tried removing the `OnAppearing` method in the page, but NRE still occurs.

### Workaround

none identified.

**Comment by HectorMeneses333** [comment]
Hi all,

Just to give you some extra info, we would like to start using Forms in a couple of large iOS and Android Xamarin native apps. We know that Forms can greatly help us simplify the workload we have on these apps but his issue is definitely a showstopper on our project. We hope you guys can raise the priority of this issue and get it fixed soon. 

Best,
Hector

**Comment by jalbertSyncroTech** [comment]
@conceptdev @samhouts @hartez  I've tracked down the cause of the issue to this pull request "Unhook Page from embedded Application on disposal" https://github.com/xamarin/Xamarin.Forms/pull/6735

The bug is that when an Application sub-class implements IDisposable, the above change incorrectly disposes the Application whenever the MainPage is set.

The chain of events is as follow:
1. The iOS FormsApplicationDelegate calls UpdateMainPage() which disposes the Platform via the RootViewController PlatformRenderer https://github.com/xamarin/Xamarin.Forms/blob/c1bdf538eabedbc27cd61c9bcae673e363b3d6e1/Xamarin.Forms.Platform.iOS/FormsApplicationDelegate.cs#L179
2. The Platform disposes the PlatformRenderer https://github.com/xamarin/Xamarin.Forms/blob/c1bdf538eabedbc27cd61c9bcae673e363b3d6e1/Xamarin.Forms.Platform.iOS/Platform.cs#L72
3. The PlatformRenderer, as part of it's Dispose() calls Platform.CleanUpPages() https://github.com/xamarin/Xamarin.Forms/blob/c1bdf538eabedbc27cd61c9bcae673e363b3d6e1/Xamarin.Forms.Platform.iOS/PlatformRenderer.cs#L110
4. Platform CleanUpPages() disposes its Parent https://github.com/xamarin/Xamarin.Forms/blob/c1bdf538eabedbc27cd61c9bcae673e363b3d6e1/Xamarin.Forms.Platform.iOS/Platform.cs#L597
5. [BUG] Since the Page is the MainPage, and it's Parent is the Application, the entire Application gets Disposed.

For the native2forms, the EmbeddedApplication is disposable, which triggers this bug by setting NavigationProxy to null.
1. The native2forms sample uses CreateViewController() to embed the Xamarin.Forms page, which creates a new EmbeddedApplication https://github.com/xamarin/Xamarin.Forms/blob/2c9b7fd38289403f388b2d0027eaeb8d75319364/Xamarin.Forms.Platform.iOS/PageExtensions.cs#L15
2. The EmbeddedApplication implements IDisposable https://github.com/xamarin/Xamarin.Forms/blob/2c9b7fd38289403f388b2d0027eaeb8d75319364/Xamarin.Forms.Platform.iOS/PageExtensions.cs#L28
3. And Application.CleanUp() nulls out the NavigationProxy https://github.com/xamarin/Xamarin.Forms/blob/1532d46f5fb6f219594015b80f8e5a2476512882/Xamarin.Forms.Core/Application.cs#L364

**Comment by jalbertSyncroTech** [comment]
@conceptdev @HectorMeneses333 

Follow up to my last comment, I've discovered a work-around as well
1. Create or update the Application sub-class to override OnPropertyChanged()
2. Before calling the base method, check is the propertyName being set is "MainPage" and save the result to a member variable.
3. In the Application sub-class Dispose() method, only proceed if the above check is false, a.k.a. only dispose if not currently setting the MainPage.
4. After calling the base method in OnPropertyChanged(), set the member variable back to false.

For native2forms, you'll need to duplicate the internal EmbeddedApplication sub-class, and also set the embedded Xamarin.Forms page's Parent *before* calling CreateViewController()
https://github.com/xamarin/Xamarin.Forms/blob/2c9b7fd38289403f388b2d0027eaeb8d75319364/Xamarin.Forms.Platform.iOS/PageExtensions.cs#L24

**Comment by HectorMeneses333** [comment]
Hi @jalbertSyncroTech,

So basically this is a workaround that needs to be implemented inside of Xamarin.Forms... Is there a workaround that we can implement that does not require changing the insides of Xamarin.Forms?

Thanks!

**Comment by jalbertSyncroTech** [comment]
@HectorMeneses333 no, the workaround is intended for your application's own sub-class of the Xamarin.Forms Application class.

**Comment by HectorMeneses333** [comment]
@jalbertSyncroTech Thanks! I'm not sure how to implement that ... so I'll wait for Microsoft to come up with an official fix. 

I know that if I comment out this call:
topView.Dispose()
in AppNavigationController.cs, then the issue disappears but it also creates a memory leak.

@samhouts, do you know when is a fix for this issue going to be implemented?

Thanks!
